internal/http/services/opencloudmesh/ocmd/shares.go

```diff
@@ -263,42 +249,82 @@ func getOCMShareType(t string) ocm.ShareType {
 func getAndResolveProtocols(p Protocols, r *http.Request) ([]*ocm.Protocol, error) {
 	protos := make([]*ocm.Protocol, 0, len(p))
 	for _, data := range p {
+		var uri string
 		ocmProto := data.ToOCMProtocol()
-		if GetProtocolName(data) == "webdav" && ocmProto.GetWebdavOptions().Uri == "" {
-			// This is an OCM 1.0 payload with only webdav: we need to resolve the remote URL
-			remoteRoot, err := discoverOcmWebdavRoot(r)
-			if err != nil {
-				return nil, err
+		protocolName := GetProtocolName(data)
+		switch protocolName {
+		case "webdav":
+			uri = ocmProto.GetWebdavOptions().Uri
+			reqs := ocmProto.GetWebdavOptions().Requirements
+			if len(reqs) > 0 {
+				// we currently do not support any kind of requirement
+				return nil, errtypes.BadRequest(fmt.Sprintf("incoming OCM share with requirements %+v not supported at this endpoint", reqs))
 			}
-			ocmProto.GetWebdavOptions().Uri = filepath.Join(remoteRoot, ocmProto.GetWebdavOptions().SharedSecret)
+		case "webapp":
+			uri = ocmProto.GetWebappOptions().Uri
+		}
+
+		// Irrespective from the presence of a full `uri` in the payload (deprecated), validate the
+		// remote is an OCM server and resolve the remote root
+		// yet skip this if the remote is localhost (for integration tests)
+		if strings.Contains(uri, "localhost") {
+			protos = append(protos, ocmProto)
+			continue
+		}
+		remoteRoot, err := discoverOcmRoot(r, protocolName)
+		if err != nil {
+			return nil, err
+		}
+
+		uri, _ = url.JoinPath(remoteRoot, uri[strings.LastIndex(uri, "/")+1:])
+		switch protocolName {
+		case "webdav":
+			ocmProto.GetWebdavOptions().Uri = uri
+		case "webapp":
+			ocmProto.GetWebappOptions().Uri = uri
 		}
 		protos = append(protos, ocmProto)
 	}
+
 	return protos, nil
 }
 
-func discoverOcmWebdavRoot(r *http.Request) (string, error) {
-	// implements the OCM discovery logic to fetch the WebDAV root at the remote host that sent the share, see
+func discoverOcmRoot(r *http.Request, proto string) (string, error) {
+	// implements the OCM discovery logic to fetch the root at the remote host that sent the share for the given proto, see
 	// https://cs3org.github.io/OCM-API/docs.html?branch=v1.1.0&repo=OCM-API&user=cs3org#/paths/~1ocm-provider/get
 	ctx := r.Context()
 	log := appctx.GetLogger(ctx)
-	log.Debug().Str("sender", r.Host).Msg("received OCM 1.0 share, attempting to discover sender endpoint")
+
+	// assume the sender host is either given in the usual reverse proxy headers or as RemoteAddr, and that the
+	// remote end listens on https regardless if the incoming connection got its TLS terminated upstream of us
+	senderURL := r.Header.Get("X-Real-Ip")
```

Comment:
I'm still not sure why we need this instead of just reading it from the request body (eventually signed). Because you can fake headers as easily as you can fake the body data, so I would think you may just as well just take whatever the initiator declares, which makes the behaviour clear, instead of relying on a bunch of headers and as a fallback the real remote addr

Reply:
You know what? I completely agree! And for two reasons:
1) Security: I thought about it on Friday after talking to you, after all the headers or the `RemoteAddr` are not more trustworthy than the payload, and eventually a signed payload will be trustworthy.
2) Functionality: for deployments such as CERNBox, with reverse proxies in between, both the `RemoteAddr` and any eventual header are simply wrong.
Therefore, I'll change the implementation to use the given `URI`, and fall back to the `RemoteAddr` (and the `X-Real-IP` header as a 2nd fall back?) only if missing (case of Nextcloud). Plus will submit a PR to the OCM specs, to "un-deprecate" the full URI in the payload.